### DISCUSSION TITLE
uvicorn can't reload when i use apscheduler and fastapi together

### DISCUSSION THREAD

**Discussion by fenghaoyu** [fastapi/fastapi#7661] [Questions] [answered]
when i use apscheduler and fastapi together like this:
```
def get_app():
    app = FastAPI()
    return app

def apscheduler():
    scheduler = AsyncIOScheduler()
    scheduler.add_job(my_jobs.tick, 'cron', hour=2, minute=0)
    deadline = datetime.datetime.now().replace(microsecond=0) + datetime.timedelta(seconds=5)
    scheduler.add_job(my_jobs.tick, 'date', run_date=deadline)
    try:
        scheduler.start()
    except Exception as e:
        scheduler.shutdown()

apscheduler()
app = get_app()
```
uvicorn can't reload automaticly like before
here is my start command:
```
uvicorn main:app --reload
```
Thank you very much if anyone can help.

**Comment by fenghaoyu** [comment]
forget to say, that it can work on windows correctly but not on linux.

**Comment by phy25** [comment]
Sound like this is more of a question suitable for uvicorn or apscheuler, if it works by removing the apscheduler part.

**Comment by selimb** [comment]
what do you mean by "can't reload"? Does it fail, or does it somehow not detect file changes anymore? What's the error message?

**Comment by fenghaoyu** [comment]
it can detect file changes, but can't reload.
the app will shut down, but can't start up.
apscheduler still keep running.
i think maybe the apscheduler block the app start up.
Strangely it works fine under windows.

**Comment by selimb** [comment]
Might be because apscheduler never actually shuts down? Try shutting it down in the [shutdown event](https://fastapi.tiangolo.com/advanced/events/#shutdown-event).

**Comment by fenghaoyu** [comment]
i have done it before,but no help.
btw,do you have any idea about getting full response json data and request json data?


------------------ Original ------------------

**Comment by fenghaoyu** [comment]
thank you guys anyway.

**Comment by tiangolo** [comment]
@fenghaoyu `--reload` is used only for development and it can have strange interactions as it's constantly reloading the code as the files change. If it's not working correctly with the other library you are using, it's probably best for you to not use `--reload` with it.

Thanks everyone for the help here! :cake: :bowing_man: 

And thanks @fenghaoyu for coming back to close the issue.

**Comment by anne17** [comment, marked as answer]
For me it worked to shut down the scheduler in the shutdown event as suggested by @selimb. I am not sure at all if this is best practice, but I store the scheduler in `app.state` in the startup event and then I can access it later:

```python
@app.on_event("startup")
async def start_scheduler():
    scheduler = AsyncIOScheduler(timezone="Europe/Stockholm")
    scheduler.add_job(...)
    scheduler.start()
    app.state.scheduler = scheduler


@app.on_event("shutdown")
async def stop_scheduler():
    app.state.scheduler.shutdown(wait=False)
```

Now uvicorn's `--reload` works as it should.

**Comment by selimb** [comment]
@anne17 That's a perfect use of `app.state` :) IMO this is a better practice than initializing everything at import time in the module scope, as recommended in the docs.
